app/pages/lab/workflow.jsx

```diff
@@ -0,0 +1,1042 @@
+        return this.context.router.push(`/lab/${this.props.project.id}`);
+    }).catch(error => {
+        return this.setState({deletionError: error});
+      }).then(() => {
+        if (this.isMounted()) {
+          return this.setState({deletionInProgress: false});
+        }
+      });
+    }
+  }
+
+  handleViewClick() {
+    return setTimeout(() => {
+      return this.setState({forceReloader: this.state.forceReloader + 1});
+    });
+  }
+
+  updateFirstTask() {
+    if (!(this.props.workflow.first_task in this.props.workflow.tasks)) {
+      let left;
+      return this.props.workflow.update({first_task: (left = Object.keys(this.props.workflow.tasks)[0]) != null ? left : ''});
+    }
+  }
+
+  prepareStepForRemoval(taskKey) {
+    const steps = [...this.props.workflow.steps];
+    const stepIndex = this.props.workflow.steps.findIndex(step => {
+      const { taskKeys } = step[1];
+      return taskKeys.includes(taskKey);
+    });
+    // Remove the step
+    steps.splice(stepIndex, 1);
+
+    return { steps, stepIndex };
+  }
+
+  deleteStepAndTask(taskKey) {
+    const changes = {};
+    const { steps, stepIndex } = this.prepareStepForRemoval(taskKey);
+
+    // Logic branch: does the Task exist in a Step?
+    if (stepIndex >= 0) {
+      // If the Task exists in a Step, then delete ALL the Tasks in that Step
+      this.props.workflow.steps[stepIndex][1].taskKeys.forEach(taskKey2 => {
+        if (this.props.workflow.tasks[taskKey2]) {
+          return changes[`tasks.${taskKey2}`] = undefined;
+        }
+      });
+
+      // Now remove the Step
+      changes.steps = steps;
+
+    } else {
+      // If the Task doesn't exist in a Step, just delete it from the Tasks list
+      changes[`tasks.${taskKey}`] = undefined;
+    }
+
+    if (changes.steps?.length === 0) {
+      // If no more steps, remove the classifier version 2.0 designation
+      changes["configuration.classifier_version"] = undefined;
+    }
+
+    return this.props.workflow.update(changes);
+  }
+
+  handleTaskDelete(taskKey, e) {
+    const shortcut = this.props.workflow.tasks[taskKey].unlinkedTask;
+    if (e.shiftKey || confirm('Really delete this task?')) {
+      if (this.canUseTask(this.props.project, "transcription-task")) {
+        this.deleteStepAndTask(taskKey);
+      } else {
+        const changes = {};
+        if (shortcut) {
+          changes[`tasks.${shortcut}`] = undefined;
+        }
+        changes[`tasks.${taskKey}`] = undefined;
+        this.props.workflow.update(changes);
+      }
+
+      return this.updateFirstTask();
+    }
+  }
+}
+
+EditWorkflowPage.contextTypes = {
+  router: PropTypes.object.isRequired
+};
+
+EditWorkflowPage.defaultProps = {
+  workflow: null,
+  workflowActions: workflowActions
+};
+
+export default function EditWorkflowPageWrapper (props) {
+  const params = props.params || {
+    workflowID: ''
+  }
+
+  return <PromiseRenderer promise={apiClient.type('workflows').get(params.workflowID, {})}>{workflow => {
```

Comment:
This `PromiseRenderer` should be replaced by a `useState` hook that stores the workflow in component state after fetching it from the API. Otherwise, you’re making network requests here even after you’ve already fetched the workflow. 

That same comment holds for the promises that fetch the tutorials and mini courses.

Reply:
I’ve included an example of this in #6219, when the project editor loads the project.

Reply:
#6226 fetches the workflow with a hook, which should only run when the workflow ID changes.